### PR TITLE
feat: app layer poc

### PR DESCRIPTION
**Pull request by alexp3y**
WIP and meant for illustrative purposes. Demonstrates adding inversifyJS to manage and organize app-layer services using an IoC container. Implements balance fetching and demos some utxo fetching logic. 

Services are made available to mobile components via react context.

@kyranjamie @pete-watters

### CODE REVIEW

**Comment by kyranjamie** [comment]
Thanks for putting together Alex, this is great. A really smart outline of how we can structure tonnes of complex logic that is currently very coupled to the React frontend. 

This service architecture set up is something you can predominantly own, so if you're adamant this is the way to go, I'm supportive. Some questions on the choices though:

Invesifyjs implements DI/IoC which are object-oriented patterns. The rest of our codebase largely follows functional patterns: the pure functions of the domain layer, our state management etc. React itself is functional by nature.

While the services layer is encapsulated, and doesn't necessarily expose the OO patterns externally, I wonder if using a mix of patterns across our stack might confuse developers. Here we use classes and DI with inversify, but there we use pure fns and avoid classes etc.

Do you think it would be possible get the architectural benefits of a DI pattern like this, but using a more functional style with HOFs? Skipping inversify, which does seems to be out of maintenance. What are the trade offs of a DI library like this, versus implementing the same thing manually?

These are pretty similar in as far as testability, and ability to control/override dependencies going in to a service. In the former, for example, we take care of instantiating the service, something Inversify would take care of in latter example, but without having to wire up a DI framework.

```ts
// Functional
function createBalanceService({ utxoService }) {}
// Inversify
class BalanceService {
  constructor(public utxoService) {}
}
```

Curious to hear your thoughts. And interested to get input from the rest of team when you're ready.

**Comment by alexp3y** [comment]
Thanks for the feedback @kyranjamie. Definitely not adamant about this pattern or the Inversify framework specifically. The driving force behind the design here is dependency management. IoC container frameworks are generally written using class annotations and, as you mentioned, its usage would be largely isolated here in the service layer, which is why I went with an object-oriented pattern.

I do agree though that mixing in OOP to a largely functional codebase is not ideal for the exact reasons you mentioned. 

Your implementation of `BalanceService` in [API Layer thoughts](https://www.notion.so/trustmachines/API-layer-concerns-11ed74b694c78062b697f6ac6a80df76?pvs=4) does a great job of demonstrating how we can approach service objects using functional patterns.

I could definitely be missing something obvious but the problem I keep coming back to is dependency management.

I've tried to restate the problem a bit more concretely along with some thoughts on how we could approach it using a more functional style [here](https://www.notion.so/p3y/Dependency-Management-1214e431827780c69e63e3eab3677ae6?pvs=4).

**Comment by alexp3y** [comment]
@kyranjamie Here’s another iteration that I think strikes a good balance between the functional style we’ve discussed and the necessary object-oriented (OOP) aspects. I considered approaches like function injection and HOF, but ultimately found that these would require sacrificing some key dependency injection (DI) functionality that we need.

I’m still using InversifyJS, but I’ve worked to minimize reliance on classes as much as possible. By following basic conventions—keeping objects stateless aside from their injected dependencies and extracting logic into pure or utility functions wherever possible—I think we can leverage the benefits of a DI/IoC framework while keeping the codebase aligned with a functional style.

**Comment by pete-watters** [comment]
Nice work @alexp3y 👍 this is coming along nicely. 

I think it's OK to have a slight mixture of OOP/ functional code when we need to. We can endeavour to use functional in the app itself but I think it's OK in isolated cases to use the best fit to make things easier to manage

**Comment by alexp3y** [comment]
Removed the services context in favor of a simplified initialization function. Moves the responsibility of defining the app service API to the services package itself. The DI container(s) and service resolver functions are held in module-scope and can be directly imported by components / hooks.

**Comment by alexp3y** [comment]
Got this working without classes or decorators. All services can now be represented by HOFs - e.g. `createBalanceService(dep1, dep2)`, with dependencies resolved by the DI container.

The HOF and its parameters need to be registered by symbol using `.toDynamicValue`, which requires a bit more boilerplate in `inversify.config.ts`, but the service implementations are now much cleaner and in-line with the functional codebase style. All previous DI functionality is preserved.

File: packages/services/src/inversify.config.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import { DefaultNetworkConfigurations, defaultNetworksKeyedById } from '@leather.io/models';
+import {
+  AccountBalanceService,
+  createAccountBalanceService,
+} from 'balances/account-balance.service';
+import { createStxBalanceService, StxBalanceService } from 'balances/stacks/stx-balance.service';
+import { QueryClientService } from 'cache/query-client.service';
+
+// Symbols for Service Bindings
+const TYPES = {
```
Comment (kyranjamie) [inline packages/services/src/inversify.config.ts:21]: I know docs call it this way but can we call it `Services` instead

File: packages/services/src/inversify.config.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  QueryClientService: Symbol.for('QueryClientService'),
+  NetworkSettingsService: Symbol.for('NetworkSettingsService'),
+  BlockbookApiClient: Symbol.for('BlockbookApiClient'),
+  MempoolSpaceApiClient: Symbol.for('MempoolSpaceApiClient'),
+  AccountBalanceService: Symbol.for('AccountBalanceService'),
+  BtcBalanceService: Symbol.for('BtcBalanceService'),
+  StxBalanceService: Symbol.for('StxBalanceService'),
+  UtxoSerivce: Symbol.for('UtxoSerivce'),
```
Comment (kyranjamie) [inline packages/services/src/inversify.config.ts:29]: typo

**Comment by alexp3y** [comment]
Removed all class and annotation-related packages and plugins. Added example unit tests.

**Comment by fbwoolf** [comment]
Really love the changes being proposed here, I think this approach is going to make things so much easier to work with moving forward. 👍

**Pull request closed without merging**